### PR TITLE
Add support for MiniDSP 2x4HD

### PR DESCRIPTION
**Pull request by dennisfrett**

### CODE REVIEW

File: MiniDSP.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ Contact information
+ -------------------
+ Kristian Lauszus, TKJ Electronics
+ Web      :  http://www.tkjelectronics.com
+ e-mail   :  kristianl@tkjelectronics.com
+ */
+
+#include "MiniDSP.h"
+
+namespace {
```
Comment (Lauszus) [inline MiniDSP.cpp:17]: There is no reason to use a empty namespace.

File: MiniDSP.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+uint8_t RequestStatusOutputCommand[] = {0x05, 0xFF, 0xDA, 0x02};
```
Comment (Lauszus) [inline MiniDSP.cpp:18]: Please move this inside `RequestStatus`.

File: MiniDSP.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+uint8_t StatusInputCommand[]{0x05, 0xFF, 0xDA};
```
Comment (Lauszus) [inline MiniDSP.cpp:19]: Please move this inside `ParseHIDData`.

File: MiniDSP.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+// Returns first byte of the sum of given bytes.
+uint8_t Checksum(const uint8_t *data, uint16_t nbytes) {
```
Comment (Lauszus) [inline MiniDSP.cpp:22]: Please make this a private function in the class.

Comment (Lauszus) [reply MiniDSP.cpp:22]: Also `nbytes` can be a `uint8_t`

File: MiniDSP.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  int sum = 0;
```
Comment (Lauszus) [inline MiniDSP.cpp]: Please change sum to a `uint16_t`

File: MiniDSP.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  for (int i = 0; i < nbytes; i++) {
```
Comment (Lauszus) [inline MiniDSP.cpp]: Please change `i` to a `uint8_t`

File: MiniDSP.cpp [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+}
+} // namespace
+
+void MiniDSP::ParseHIDData(USBHID *hid __attribute__((unused)),
+                           bool is_rpt_id __attribute__((unused)), uint8_t len,
+                           uint8_t *buf) {
+
+  // Only care about valid data for the MiniDSP 2x4HD.
+  if (HIDUniversal::VID != MINIDSP_VID || HIDUniversal::PID != MINIDSP_PID ||
+      len <= 2 || buf == nullptr) {
```
Comment (Lauszus) [inline MiniDSP.cpp:38]: This should be `len < 4` or you will have a buffer over-read if `len` is 3, as `StatusInputCommand` is 3 bytes long i.e. you will be reading parsed the `buf`.

File: MiniDSP.cpp [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+    const auto volumeUpdated = newVolume != volume;
+    const auto isMutedUpdated = newIsMuted != isMuted;
+
+    // Update status.
+    volume = newVolume;
+    isMuted = newIsMuted;
+
+    // Call callbacks.
+    if (volumeChangeCallback != nullptr && volumeUpdated) {
```
Comment (Lauszus) [inline MiniDSP.cpp:59]: Why not just do:

```C++
if (volumeChangeCallback != nullptr && newVolume != volume)
  volumeChangeCallback(volume);
volume = newVolume;
```

Same goes with the other callback.

File: MiniDSP.cpp [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+      mutedChangeCallback(isMuted);
+    }
+  }
+};
+
+uint8_t MiniDSP::OnInitSuccessful() {
+  // Verify we're actually connected to the MiniDSP 2x4HD.
+  if (HIDUniversal::VID != MINIDSP_VID || HIDUniversal::PID != MINIDSP_PID) {
+    return 0;
+  }
```
Comment (Lauszus) [inline MiniDSP.cpp:73]: Please remove brackets to follow the coding standard of the library. Also could you also please 4 or 8 spaces as indentation?

File: MiniDSP.cpp [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+  return 0;
+};
+
+void MiniDSP::SendCommand(uint8_t *command, uint16_t command_length) const {
+  // Only send command if we're actually connected to the MiniDSP 2x4HD.
+  if (HIDUniversal::VID != MINIDSP_VID || HIDUniversal::PID != MINIDSP_PID) {
+    return;
+  }
```
Comment (Lauszus) [inline MiniDSP.cpp:89]: No reason to check VID and PID, as it has already been checked before calling `RequestStatus`.

File: MiniDSP.cpp [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  RequestStatus();
+
+  if (onInitCallback != nullptr) {
+    onInitCallback();
+  }
+
+  return 0;
+};
+
+void MiniDSP::SendCommand(uint8_t *command, uint16_t command_length) const {
```
Comment (Lauszus) [inline MiniDSP.cpp:85]: `command_length` can be uint8_t

File: MiniDSP.cpp [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+  // [ length (command + checksum byte) ] [ command ] [ checksum ] [ OxFF... ]
+
+  // MiniDSP expects 64 byte messages.
+  uint8_t buf[64];
+
+  // Set length, including checksum byte.
+  buf[0] = command_length + 1;
+
+  // Copy actual command.
+  memcpy(&buf[1], command, command_length);
```
Comment (Lauszus) [inline MiniDSP.cpp:101]: Maybe add a check to make sure that `command_length` is not more than 63.

File: MiniDSP.h [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+ Contact information
+ -------------------
+ Kristian Lauszus, TKJ Electronics
+ Web      :  http://www.tkjelectronics.com
+ e-mail   :  kristianl@tkjelectronics.com
+ */
+
+#pragma once
+
+#include "controllerEnums.h"
```
Comment (Lauszus) [inline MiniDSP.h:17]: `#include "controllerEnums.h"` can be removed.

File: MiniDSP.h [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+  bool connected() {
+    return HIDUniversal::isReady() && HIDUniversal::VID == MINIDSP_VID &&
+           HIDUniversal::PID == MINIDSP_PID;
+  };
+
+  /**
+   * Used to call your own function when the device is successfully initialized.
+   * @param func Function to call.
+   */
+  void SetOnInitCallback(void (*func)(void)) { onInitCallback = func; };
```
Comment (Lauszus) [inline MiniDSP.h:54]: Please rename to `attachOnInit`, so it is the same as all the other drivers fx: https://github.com/felis/USB_Host_Shield_2.0/blob/master/PS5USB.h#L53-L59

File: MiniDSP.h [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+  /**
+   * Used to call your own function when the volume has changed.
+   * The volume is passed as an unsigned integer that represents twice the -dB
+   * value. Example: 19 represents -9.5dB.
+   * @param func Function to call.
+   */
+  void SetVolumeChangeCallback(void (*func)(uint8_t)) {
```
Comment (Lauszus) [inline MiniDSP.h:62]: Hmm could you change this to `attachVolumeChangeCallback` as well.

File: MiniDSP.h [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+    volumeChangeCallback = func;
+  }
+
+  /**
+   * Used to call your own function when the muted status has changed.
+   * The muted status is passed as a boolean. True means muted, false means
+   * unmuted.
+   * @param func Function to call.
+   */
+  void SetMutedChangeCallback(void (*func)(bool)) {
```
Comment (Lauszus) [inline MiniDSP.h:72]: Please change to `attachMutedChangeCallback`

File: MiniDSP.h [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    mutedChangeCallback = func;
+  }
+
+  /**
+   * Retrieve the current volume of the MiniDSP.
+   * The volume is passed as an unsigned integer that represents twice the -dB
+   * value. Example: 19 represents -9.5dB.
+   * @return Current volume.
+   */
+  int GetVolume() const { return volume; }
```
Comment (Lauszus) [inline MiniDSP.h:82]: Please change to `getVolume()`

Comment (Lauszus) [reply MiniDSP.h:82]: Maybe also add a function called `float getVolumeDB` as well?

File: MiniDSP.h [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+
+  /**
+   * Retrieve the current muted status of the MiniDSP
+   * @return `true` if the device is muted, `false` otherwise.
+   */
+  bool IsMuted() const { return isMuted; }
```
Comment (Lauszus) [inline MiniDSP.h:88]: Please rename to `isMuted`

File: examples/MiniDSP/MiniDSP.ino [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+// Satisfy the IDE, which needs to see the include statment in the ino too.
+#ifdef dobogusinclude
+#include <spi4teensy3.h>
+#endif
+#include <SPI.h>
+
+USB Usb;
+MiniDSP MiniDSP(&Usb);
+
+void OnMiniDSPConnected() { Serial.println("MiniDSP connected"); }
```
Comment (Lauszus) [inline examples/MiniDSP/MiniDSP.ino:18]: Please do not have all on one line.

File: examples/MiniDSP/MiniDSP.ino [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+  }
+  Serial.println(F("\r\nMiniDSP 2x4HD Library Started"));
+
+  // Register callbacks.
+  MiniDSP.SetOnInitCallback(&OnMiniDSPConnected);
+  MiniDSP.SetVolumeChangeCallback(&OnVolumeChange);
+  MiniDSP.SetMutedChangeCallback(&OnMutedChange);
+}
+
+void loop() { Usb.Task(); }
```
Comment (Lauszus) [inline examples/MiniDSP/MiniDSP.ino:48]: Please use 3 lines.

File: examples/MiniDSP/MiniDSP.ino [hunk 21]
```diff
@@ -0,0 +1,48 @@
+/*
+ Example sketch for the Playstation Buzz library - developed by Kristian Lauszus
+ For more information visit my blog: http://blog.tkjelectronics.dk/ or
+ send me an e-mail:  kristianl@tkjelectronics.com
+ */
```
Comment (Lauszus) [inline examples/MiniDSP/MiniDSP.ino:5]: Please update this header text.

**Review by Lauszus — CHANGES_REQUESTED** [review]

File: MiniDSP.cpp [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+
+#include "MiniDSP.h"
+
+namespace {
+uint8_t RequestStatusOutputCommand[] = {0x05, 0xFF, 0xDA, 0x02};
+uint8_t StatusInputCommand[]{0x05, 0xFF, 0xDA};
+
+// Returns first byte of the sum of given bytes.
+uint8_t Checksum(const uint8_t *data, uint16_t nbytes) {
+  uint64_t sum = 0;
```
Comment (Lauszus) [inline MiniDSP.cpp:23]: Please change sum to a `uint16_t`

File: MiniDSP.cpp [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+  for (uint16_t i = 0; i < nbytes; i++) {
```
Comment (Lauszus) [inline MiniDSP.cpp:24]: Please change `i` to a `uint8_t`

**Comment by Lauszus** [comment]
@dennisfrett thanks a lot for the pull request. I have left a bunch of comments. Could you take a look at them?

File: MiniDSP.cpp (hunk 8 above)
Comment (dennisfrett) [reply MiniDSP.cpp]: I would rather the state be updated before the callbacks are invoked. The volume-callback can use the new muted-state, if it was updated.

**Comment by xxxajk** [comment]
@Lauszus I can always auto-format after @dennisfrett is done. Perhaps we need to enforce this better and post a wiki page? I have the entire list of settings.

@dennisfrett indents are 8 spaces, plus there are a few other code style things to watch for. Why? makes it easy to read. Basically if you follow the Linux Kernel code style, you'll have much easier to read code.

**Comment by dennisfrett** [comment]
Yeah, in my last update I tried to match the code style more. I'm auto-formatting with Clang-format. Maybe you could push a `.clang-format` settings file for the tool to pick up. Which formatting tool do you guys use? 

For style-choices like camelCase function names, a minter tool could be used to enforce this. 

Let me know if there is anything else I can do!

**Comment by xxxajk** [comment]
@dennisfrett Can you import NetBeans settings? That's what (at least I) use.

**Comment by dennisfrett** [comment]
> @dennisfrett Can you import NetBeans settings? That's what (at least I) use.

I'm using VSCode here. But if you guys can re-format it with your settings completely, that's perfectly fine for me. 

The basic implementation of the driver is done. I might extend it later if I need more functionality or if other people want certain features from it.

Is there anything else that you guys want me to do / change before it can be merged in? 

Thanks for the implementation help & review!

**Comment by xxxajk** [comment]
... @dennisfrett  ... Here's the link to the NetBeans settings...
[nbsettings.zip](https://github.com/felis/UHS30/blob/master/nbsettings.zip)
In NetBeans you don't unzip the file, however, if you DO unzip, there's an XML file that has the settings and they are quite human readable!

The ones that you want are for C, C++ 
One of the files that you would want to look at is 
`config/Editors/Preferences/org-netbeans-modules-editor-settings-CustomPreferences.xml`
This is considered the "base"
From there, you just look for the various .properties files.
the ones you want are at 
`config/Preferences/org/netbeans/modules/cnd/editor`
From there, just look at the various settings. 
NOTE: Defaults are in extra directories, so for example C++
You start with the base file above
Then replace any changed values from
`CPP_CodeStyles/Default.properties`
and then you end up with all the settings.

Easier to use NetBeans and just import the settings, though ...

**Comment by dennisfrett** [comment]
@xxxajk , reformatted the files with your settings, could you verify if it's correct? Thanks!

**Comment by xxxajk** [comment]
@dennisfrett  much better!

**Comment by Lauszus** [comment]
@dennisfrett looks great! Actually the ino files should only be 2 spaces, as that is what the Arduino editor uses :/

**Comment by xxxajk** [comment]
Actually, you can set that to 8. It's in the settings, but the silly excuse for an IDE does a lousy job.

File: MiniDSP.h [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+ Foundation and appearing in the file GPL2.TXT included in the packaging of
+ this file. Please note that GPL2 Section 2[b] requires that all works based
+ on this software must also be made publicly available under the terms of
+ the GPL2 ("Copyleft").
+ Contact information
+ -------------------
+ Kristian Lauszus, TKJ Electronics
+ Web      :  http://www.tkjelectronics.com
+ e-mail   :  kristianl@tkjelectronics.com
+ */
```
Comment (Lauszus) [inline MiniDSP.h:13]: You can also update the copyright header to add yourself  :)


Also can you use this one instead: https://github.com/felis/USB_Host_Shield_2.0/blob/master/XBOXONESBT.h#L1-L16, as I do not use that email anymore.

File: MiniDSP.cpp [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+ Foundation and appearing in the file GPL2.TXT included in the packaging of
+ this file. Please note that GPL2 Section 2[b] requires that all works based
+ on this software must also be made publicly available under the terms of
+ the GPL2 ("Copyleft").
+ Contact information
+ -------------------
+ Kristian Lauszus, TKJ Electronics
+ Web      :  http://www.tkjelectronics.com
+ e-mail   :  kristianl@tkjelectronics.com
+ */
```
Comment (Lauszus) [inline MiniDSP.cpp:13]: Same with this copyright header.

**Comment by Lauszus** [comment]
@xxxajk yes, but Arduino is meant for beginning, so lets not scare them away with telling them to change their editor settings etc. Lets just go with what ever is standard for the ino example.

**Comment by dennisfrett** [comment]
I did some final fixes, let me know if anything else is needed before this can be merged.

**Review by Lauszus — APPROVED** [review]

**Comment by Lauszus** [comment]
Thanks again for the PR. I've just merged it.

**Pull request merged**